File: Week 02/id_118/C++/LeetCode_17_118.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <unordered_map>

using namespace std;


class Solution {
private:

    // mapping letter and number
    const string letterMap[10] = {
            " ",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
    };

    // res to store the result
    vector<string> res;

    // s is the current string from [0,...,index-1]
    // at the current index level, just look for digits[index] => to get digits[0,...,index]
    void findCombination(const string &digits, int index, const string &s) {

        // terminator
        if (index == digits.size()) {
            // here s is one solution
            // save s to res
            res.push_back(s);
            return;
        }

        // get the char of the digit
        char c = digits[index];
        // make sure the char is between 0 and 9, also 1 doesn't have any char
        assert(c >= '0' && c <= '9' && c != '1');
        // get letters based on the current number
        string letters = letterMap[c - '0'];
        for (int i = 0; i < letters.size(); ++i) {
            // drill down
            findCombination(digits, index + 1, s + letters[i]);
        }

        return;
    }

public:
    vector<string> letterCombinations(string digits) {
        // initialization
        res.clear();

        // edge case, empty string input => empty output, not empty string output
        if (digits == "") {
            return res;
        }
        findCombination(digits, 0, "");
        return res;
    }
};

//leetcode submit region end(Prohibit modification and deletion)
int main() {
    vector<string> res = Solution().letterCombinations("4567");
    for (int i = 0; i < res.size(); ++i) {
        cout << res[i] << endl;
    }
    return 0;
};

File: Week 02/id_118/Python/LeetCode_17_118.py
from typing import List


class Solution:
    """
    This solution strictly follows the recursion template
    Time complexity is O(3^n), exponential
    """

    def __init__(self):
        """
        initiate with a result to store
        letter_map dict to store the mapping between number and letters
        """
        self.res = []
        self.letter_map = {
            0: " ",
            1: "",
            2: "abc",
            3: "def",
            4: "ghi",
            5: "jkl",
            6: "mno",
            7: "pqrs",
            8: "tuv",
            9: "wxyz"
        }

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return self.res
        self._letter_combo(0, len(digits), digits, "")
        return self.res

    def _letter_combo(self, curr_level, max_level, digits, curr_str):
        """

        :param curr_level: current digit
        :param max_level: length of the digits
        :param digits: digits
        :param curr_str: current letter string
        :return:
        """
        # for each curr_level, print out the curr_str
        print("current level is {a} : current str is {b}".format(a=curr_level,
                                                                 b=curr_str))

        # terminator
        if curr_level >= max_level:
            self.res.append(curr_str)
            return

        # process the current logic
        # get the current digit
        digit = digits[curr_level]
        # make sure the digit is valid
        assert (0 <= int(digit) <= 9) and (int(digit) != 1)

        # get the letters of the digit
        letters = self.letter_map[int(digit)]

        # drill down for each letter
        for letter in letters:
            # print out the details for each recursion
            print("digits[{a}] = {b}, letter {c} is used".format(a=curr_level,
                                                                 b=digit,
                                                                 c=letter))
            self._letter_combo(curr_level + 1, max_level, digits,
                               curr_str + letter)

        # after finishing recursion at each level, revert to the last level
        return


def test():
    sol = Solution()
    res = sol.letterCombinations("5678")
    print(res)


if __name__ == '__main__':
    test()
